presto-main/src/main/java/com/facebook/presto/execution/QueryManagerStats.java

```diff
@@ -169,15 +171,25 @@ public void stateChanged(QueryState newValue)
                     if (started) {
                         queryStopped();
                     }
-                    else {
+                    else if (queued) {
                         queryDequeued();
+                        queued = false;
                     }
                     finalQueryInfoSupplier.get()
                             .ifPresent(QueryManagerStats.this::queryFinished);
                 }
+                else if (newValue.ordinal() > QUEUED.ordinal() && newValue.ordinal() < RUNNING.ordinal()) {
```

Comment:
Correct me if I'm wrong, but where is the logic of updating `queuedQueries` when the new state is QUEUED? In such case the newValue.ordinal() is equal to QUEUED.ordinal(), also the newValue.isDone() is false. Which if branch handles this situation?

Reply:
Should we introduce `queryQueued()` method that increment queuedQueries count, and add a branch  in this method if the new state is QUEUED?

Reply:
> where is the logic of updating queuedQueries when the new state is QUEUED?

In the current implementation, this was not required since every call to trackQueryStats was incrementing queuedQueries. 

> Should we introduce queryQueued() method that increment queuedQueries count, and add a branch in this method if the new state is QUEUED?

But yes, I think this looks better, I just saw that query is queued after trackQueryStats(dispatchQuery). So we should not be incrementing the queuedQueries in trackQueryStats.